### PR TITLE
add promptflow example

### PR DESCRIPTION
**Pull request by lordlinus**
## Why are these changes needed?

This pull request introduces an example of how to incorporate Autogen within a workflow using PromptFlow. PromptFlow enables the creation, debugging, sharing, and iteration of executable flows, which can then be deployed as API endpoints. Additionally, it provides capabilities for evaluating and monitoring the flow to meet the requirements of enterprise environments.

## Checks

- [X] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [X] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [X] I've made sure all auto checks have passed.

### CODE REVIEW

**Review by ekzhu — CHANGES_REQUESTED** [review]
Thanks for the PR! A few things:

1. Can you add some illustrations of what is PromptFlow and how this application integrates PromptFlow with AutoGen. Right now the README dives directly into "Chat flow" but many users do not know what this means.
2. Can you add a blub to the Ecosystem page for PromptFlow? You can find existing examples in `website/docs/ecosystem`.

**Review by d3ssy — APPROVED** [review]
Great work! Would benefit from an explanation of the usage and role of Redis as well as alternatives to capture intermediate msgs during agent interactions.

**Comment by lordlinus** [comment]
> Thanks for the PR! A few things:
> 
> 1. Can you add some illustrations of what is PromptFlow and how this application integrates PromptFlow with AutoGen. Right now the README dives directly into "Chat flow" but many users do not know what this means.
> 2. Can you add a blurb to the Ecosystem page for PromptFlow? You can find existing examples in `website/docs/ecosystem`.

Thank you @ekzhu for your review, i have updated Readme. Can you please take a look again?

**Comment by sonichi** [comment]
@ekzhu the PR is blocked by your change request.

File: samples/apps/promptflow-autogen/flow.dag.yaml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    type: string
+    reference: ${autogen_task.output}
+    is_chat_output: true
+nodes:
+- name: autogen_task
+  type: python
+  source:
+    type: code
+    path: autogen_task.py
+  inputs:
```
Comment (krishnashed) [inline samples/apps/promptflow-autogen/flow.dag.yaml:27]: What if the user misses defining a few values in YAML, do we have proper error handling messages thrown ?

In all of these YAML files and json schema, how will the user know which variables can be skipped, and which of them need to be defined and are required ?

**Review by krishnashed — CHANGES_REQUESTED** [review]

**Comment by krishnashed** [comment]
@lordlinus are you available on Autogen's discord? I would like to understand your code better by hopping on a call if possible. I'm facing some errors executing your code on my side. Maybe it's just some environment variables not being set or something trivial.

**Review by krishnashed — CHANGES_REQUESTED** [review]
Error handling and Optional fields if any in YAML and JSON Schema

**Comment by lordlinus** [comment]
> @lordlinus are you available on Autogen's discord? I would like to understand your code better by hopping on a call if possible. I'm facing some errors executing your code on my side. Maybe it's just some environment variables not being set or something trivial.

my Discord id `ssk_sg` can you please share yours to connect?

**Review by ekzhu — APPROVED** [review]

**Review by Josephrp — APPROVED** [review]
great start to see more of these shared : 
- promptflow is great
- semantic kernel is great
- kernel memory is great

the documentation is 👌🏻very well done

File: samples/apps/promptflow-autogen/flow.dag.yaml (hunk 1 above)
Comment (sonichi) [reply samples/apps/promptflow-autogen/flow.dag.yaml:27]: Once this conversation is resolved, the PR will be merged.

Comment (lordlinus) [reply samples/apps/promptflow-autogen/flow.dag.yaml:27]: @krishnashed lets connect on Discord to discuss further

Summary: Promptflow support creating the steps using the DAG UI and is auto-converted to YAML by the framework itself - https://microsoft.github.io/promptflow/how-to-guides/quick-start.html#understand-the-flow-yaml
all the error handling for incorrect and invalid flow.dag.yaml is handled and managed by the promtpflow framework itself. User gets a visual display on where in the flow they need to correct the issue. Example screenshot below

simulating an error by referring to non-existent file:
image

image

Comment (ElliotWood) [reply samples/apps/promptflow-autogen/flow.dag.yaml:27]: Looks good , I wanted to double check the role of Redis - is it necessary? Or simply a cache example.

Secondly the output, is formatted to work as a promptflow result "output:<JSON>", you'll find it much easier to chain promptflows together if we follow that standard.

Comment (lordlinus) [reply samples/apps/promptflow-autogen/flow.dag.yaml:27]: Thanks @ElliotWood 
- Redis is just a cache example and i am planning to have a follow up on how to take those interim messages and return to the user. Autogen output generally take longer and i want to ensure we have a way to display the interim messages if required.

- In the current use-cases there is no need to have history and hence i did not want to over complicate. I am returning `ChatResult` object and getting `summary` field. You can get `chat_history` and other fields too.. need to change from `res.summary` to `res`. 
For the next PR i want to include Group chat with Human interaction and Include history and show end to end. Let me know your thoughts

image

image


image

Comment (lordlinus) [reply samples/apps/promptflow-autogen/flow.dag.yaml:27]: @krishnashed i have updated the readme as discussed and resolving this conversation as agreed. Thanks again for your time on discord.

Comment (krishnashed) [reply samples/apps/promptflow-autogen/flow.dag.yaml:27]: Thanks @lordlinus. Rest everything looks good to me

**Comment by krishnashed** [comment]
@lordlinus I see a nice to have feature which can be developed for easy installation and setup of Promptflow. 

`pip install pyautogen[promptflow]`

This can be replicated to other integrations mentioned in as well https://github.com/microsoft/autogen/issues/140

But we can take this up later too, I'm just suggesting.

File: samples/apps/promptflow-autogen/flow.dag.yaml (hunk 1 above)
Comment (ElliotWood) [reply samples/apps/promptflow-autogen/flow.dag.yaml:27]: > Thanks @ElliotWood
> 
> * Redis is just a cache example and i am planning to have a follow up on how to take those interim messages and return to the user. Autogen output generally take longer and i want to ensure we have a way to display the interim messages if required.
> * In the current use-cases there is no need to have history and hence i did not want to over complicate. I am returning `ChatResult` object and getting `summary` field. You can get `chat_history` and other fields too.. need to change from `res.summary` to `res`.
>   For the next PR i want to include Group chat with Human interaction and Include history and show end to end. Let me know your thoughts
> 
> image
> 
> image
> 
> image

No problem - you can do something like this.

```
def _chat_history_to_openai(self, chat_history: list) -> list:
        """
            takes the chat history as produced by prompt flow and formats it for use in the simulator
            promptflow uses the following format:
                chat_history = [{"inputs":{"question":"What did OpenAI announce on March 2023?"},
                                 "outputs":{"answer":"On March 2023, OpenAI announced the release of GPT-4"}}]
            simulator uses the following format used by openai chat models:
                [{"role": "human", "content": "What did OpenAI announce on March 2023?"}, 
                 {"role": "agent", "content": "On March 2023, OpenAI announced the release of GPT-4"}]
        """
        formatted_chat_history = []
        for i in chat_history:
            formatted_chat_history.append({"role": "user", "content": i["inputs"][self.question]})
            formatted_chat_history.append({"role": "assistant", "content": i["outputs"][self.answer]})
        return formatted_chat_history
```

```
def _chat_history_to_pf(self, chat_history: list) -> list:
        """
            takes the chat history as produced by openai and converts it to the prompt flow and format
            simulator uses the following format used by openai chat models:
                [{"role": "user", "content": "What did OpenAI announce on March 2023?"}, 
                 {"role": "assistant", "content": "On March 2023, OpenAI announced the release of GPT-4"}]
            promptflow uses the following format:
                chat_history = [{"inputs":{"question":"What did OpenAI announce on March 2023?"},
                                 "outputs":{"answer":"On March 2023, OpenAI announced the release of GPT-4", "context":"some context for the prior answer"}}]
        """
        chat_history = chat_history.copy()
        # filter out any messages that are not from the user or assistant
        chat_history = [message for message in chat_history if message["role"] in ["user", "assistant"]]
        pf_chat_history = []
        for i in range(0, len(chat_history)-1, 2):
            pf_chat_history.append({"inputs":{self.question: chat_history[i]["content"]},
                                    "outputs":{self.answer: chat_history[i+1]["content"]}})
        return pf_chat_history
```

**Pull request merged**